### ISSUE TITLE
pre-deployment-checklist-mandatory: sub-checks re-demand concerns that notApplicable already answered

### ISSUE THREAD

**Issue by ruleofcodedev** [Rule-Of-Code/RoC#96]
**Version:** 7.20.0
**Law:** `pre-deployment-checklist-mandatory`

## What happens

The checklist part of this law is satisfiable and we satisfied it. What remains
cannot be:

```
Authentication/authorization setup not detected
Performance benchmarking not configured
Health checks not configured
Logging not properly configured
```

Those are the same concerns as three laws that this project has already declared
not-applicable, with written reasons, and which the tool accepted:

```jsonc
// ruleofcode.config.json — laws.notApplicable
"authentication-security":  "No authentication surface: prerendered static site …"
"api-security-standards":   "No API and no deployed server: firebase.json publishes only
                             dist/apps/web/browser and declares no Functions …"
"health-check-monitoring":  "No running service to health-check: prerendered HTML on
                             static hosting, no server process, no database …"
```

So the audit accepts that this project has no authentication and nothing to
health-check — and then a different law asks for both again, in a place
`notApplicable` cannot reach. A sub-check is not addressable: there is no
`notApplicable` for "the auth clause of the checklist law".

## The checklist part works, and works well

Worth saying, because the fix is not to loosen it. Following the findings
produced a real document:

1. `PRE_DEPLOYMENT_CHECKLIST.md missing - CRITICAL violation` → wrote it
2. `has no steps for: rollback / recovery` → added them, and it was right to ask;
   the file had no recovery path
3. `contains no checklist steps` → added literal `- [ ]` items

The advice attached to (3) is the best line in this law:

> Add checklist steps covering … — **leave the boxes UNTICKED**; they are ticked
> at deployment time

That is exactly right, and it is the opposite of what a checklist law usually
incentivises. Ours is committed with 13 unticked boxes and 0 ticked.

## Why it matters

Our policy is that a law is `error` or it is declared not-applicable with a
written reason — `warning` is not a state a law should live in. This one cannot
be either: the concern genuinely exists (a deployment checklist), we meet it, and
the law still fails on clauses that three accepted waivers already answered.

It also produces a strange public artefact. Our audit page lists
`pre-deployment-checklist-(mandatory)` as unmet directly beneath a repository
containing a complete pre-deployment checklist.

## Suggested direction

Any of these resolves it; the first seems closest to how the tool already thinks:

1. **Let a sub-check defer to the law that owns the concern.** If
   `authentication-security` is not-applicable, the auth clause here is answered.
   Same for `health-check-monitoring`. The tool already has the adopter's written
   reason — it is one lookup away.
2. **Scope the clauses to project shape.** No deployed server process means no
   health endpoint and no auth surface, the same detection that already drives
   several other laws.
3. **Let the checklist answer for itself.** If the document declares a section
   as not-applicable with a reason — ours has one, headed "What this checklist
   deliberately does not contain" — treat that as the answer, since the law's
   subject is the checklist rather than the runtime.

Option 3 has a nice property: the reason ends up in the artefact a human reads
before deploying, instead of in a config file nobody opens at that moment.

**Comment by ruleofcodedev** [comment]
**Fixed — in #99, awaiting CI and merge to `develop`. Ships in the next release.**

Your option (1) is what shipped, for the reason you gave: the tool already holds the adopter's written reason, and it was one lookup away.

Each of the four clauses now defers to the law that owns its concern:

| clause | defers to |
| --- | --- |
| authentication | `authentication-security` |
| health checks | `health-check-monitoring` |
| performance benchmarking | `performance-monitoring-standards` |
| logging | `centralized-logging` |

Three boundaries held deliberately, all covered by tests:

- **The checklist clauses are untouched.** The document is this law's subject, and no waiver on another law can answer for it — delete the file and the law still fails with all four waivers in place.
- **A waiver answers its own clause and no other.**
- **An empty reason is not a waiver.** A waiver without a justification is the disarmed gate this whole tool refuses.

Resolution is by **name identity only**, not `legacyId` — that value is shared by 51 laws, and an ambiguous key must never widen a waiver silently. The registry refuses to act on one for the same reason.

We did not take option (3), the checklist declaring its own not-applicable sections, though it is the most interesting of the three. Your argument for it is the strongest — *"the reason ends up in the artefact a human reads before deploying, instead of in a config file nobody opens at that moment"* — and it deserves its own issue rather than being folded in here, because it changes what the law reads rather than which findings it suppresses. If you file it, that quote should be the opening line.

Separately: thank you for the paragraph about what the law got right. The "leave the boxes UNTICKED" advice was written deliberately against the usual incentive, and this is the first report confirming it landed.

**Comment by ruleofcodedev** [comment]
Merged to `develop` in #99. Ships in the next release.
